staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/marshal.go

```diff
@@ -20,12 +20,40 @@ import (
 	"bytes"
 	"errors"
 
+	cbor "k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct"
 	"k8s.io/apimachinery/pkg/util/json"
 )
 
 var jsTrue = []byte("true")
 var jsFalse = []byte("false")
 
+// The CBOR parsing related constants and functions below are not exported so they can be
+// easily removed at a future date when the CBOR library provides equivalent functionality.
```

Comment:
I wasn't expecting non-test changes in this PR and the addition of custom CBOR marshaling surprised me.

is there a link to the CBOR library I presume this is copying from so we can verify this is correct / doesn't drift from that code?

Reply:
I reviewed this with @sanchezl and it's not based on existing code, only the CBOR specification. The encoding of the major type in the high-order 3 bits of the initial byte is specified in https://www.rfc-editor.org/rfc/rfc8949.html#section-3-2, the major type constants in // https://www.rfc-editor.org/rfc/rfc8949.html#section-3.1, and the simple value constants are mentioned both in the RFC and in the CBOR simple value IANA registry (https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml).

Reply:
> I wasn't expecting non-test changes in this PR and the addition of custom CBOR marshaling surprised me.

@liggitt Making sure everyone is on the same page.  Is the reason you were not expecting non-test changes because there is already a JSON marshaller available and we COULD do a go-struct->json->CBOR conversion?

Reply:
> > I wasn't expecting non-test changes in this PR and the addition of custom CBOR marshaling surprised me.
> 
> @liggitt Making sure everyone is on the same page. Is the reason you were not expecting non-test changes because there is already a JSON marshaller available and we COULD do a go-struct->json->CBOR conversion?

No, I got confused by the title of the PR and thought it was just adding fuzz testing to resolve https://github.com/kubernetes/kubernetes/issues/79027. Knowing it is for the CBOR KEP makes sense why there are non-test changes 👍 

I *am* still surprised to see apparently duplicate custom low-level constants for various CBOR types and helpers (`cborType(b byte) cborMajorType`) living in these packages. Is there not a spot we can define those once and share them?

Reply:
I think @sanchezl had originally shared a single implementation between the two packages. I'd argued against introducing a CBOR syntax API, to avoid encouraging its spread and to make it painless to replace with calls to an upstream library in the future. The only diff between ./v1/marshal.go and ./v1beta1/marshal.go is the boilerplate copyright year and the package name, so I'm not convinced that we have a distinct second use case for the duplicated code yet. It's not a hill I will die on, but that's how it came to be the way it is.